[AMDGPU][clang] provide device implementation for __builtin_logb and …

choikwa opened the pull request:
…__builtin_scalbn

Clang generates library calls for __builtin_* functions which can be a problem for GPUs that cannot handle them. This patch generates call to device implementation for __builtin_logb and ldexp intrinsic for __builtin_scalbn.

choikwa committed:
[AMDGPU][clang] provide device implementation for __builtin_logb and __builtin_scalbn

Clang generates library calls for __builtin_* functions which can be a problem for GPUs that cannot handle them.
This patch generates a device implementations for __builtin_logb and __builtin_scalbn by emitting LLVM IRs.
Only emit IRs when FP exceptions are disabled and math-errno is unset.

arsenm reviewed (changes requested):
We do not and should never emit ocml calls in clang. There is also no reason to ever use scalbn for binary float types, it is identical to llvm.ldexp

choikwa commented:
The motivation came from CUDA cccl code containing __builtin_logb and __builtin_scalbn and being able to handle them.

jhuber6 commented:
> The motivation came from CUDA cccl code containing __builtin_logb and __builtin_scalbn and being able to handle them.

In my ideal world, unhandled builtins would be lowered before the linker and we'd have a working library called `libm.a` that matches the platform.

choikwa commented:
I would have preferred to do the overloads in the target specific header file, but __builtin_*  methods couldn't be overloaded. I noticed that hip math header already overloads 'double logb(double)' by calling ocml version, so I mirrored that in clang. If there is better approach, I am interested.

jhuber6 commented:
> I would have preferred to do the overloads in the target specific header file, but __builtin_* methods couldn't be overloaded. I noticed that hip math header already overloads 'double logb(double)' by calling ocml version, so I mirrored that in clang. If there is better approach, I am interested.

Long term, I think it's as I outlined in my RFC that I linked earlier.

arsenm reviewed (changes requested):
clang should never emit a call to an ocml function

choikwa commented:
> clang should never emit a call to an ocml function

Would IR body of the ocml function be a suitable replacement to emit (for now)?

I understand once we have libm/libc implementation, clang can decide between the two.

arsenm commented:
> I understand once we have libm/libc implementation, clang can decide between the two.

The behavior should be to just emit the call to the libm function. The real problem is we don't link a libm, and have a shim in a header file which isn't a system property

choikwa commented:
> > I understand once we have libm/libc implementation, clang can decide between the two.
> 
> The behavior should be to just emit the call to the libm function. The real problem is we don't link a libm, and have a shim in a header file which isn't a system property

There were some stakeholders who are trying to compile cccl on AMDGPU and were looking for a solution in clang. If this short term fix isn't desirable, is there any alternative solutions to explore? I understand long term solution keeping it as libcall and having libm, but we may be some time away from achieving that.

jhuber6 commented:
> There were some stakeholders who are trying to compile cccl on AMDGPU and were looking for a solution in clang. If this short term fix isn't desirable, is there any alternative solutions to explore? I understand long term solution keeping it as libcall and having libm, but we may be some time away from achieving that.

For now these don't seem to hit any LLVM-IR intrinsics https://godbolt.org/z/zasTefo53. So you should be able to link them with the `libm.a` I already have from the [llvm libc](https://libc.llvm.org/gpu/building.html) I already provide. That might change in the future however.

There's `libm.bc` in the install which you can cram into your compile via your favorite method. Or just use `-Xoffload-linker -lm` if you're using the new driver on HIP (I'll make it the default eventually). Alternatively just make your own `libm.o` that contains the ROCm DL implementation and link that in. And if absolutely none of that works, make a custom IR lowering pass.

jhuber6 reviewed (commented):
Are you trying to do an IR lowering? It'd probably make more sense to handle that in LLVM than clang, but I guess Matt would be the expert there.

choikwa commented:
> Are you trying to do an IR lowering? It'd probably make more sense to handle that in LLVM than clang, but I guess Matt would be the expert there.

I had considered it but it has the issue of name aliasing since it just appears as an external call in LLVM after it's emitted by clang.

choikwa commented:
gentle ping

jhuber6 commented on clang/test/CodeGenHIP/logb_scalbn.hip:
```diff
@@ -0,0 +1,32 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
+// RUN: %clang --cuda-device-only -nogpuinc -nogpulib -emit-llvm -S -o - %s | FileCheck %s
```
I feel like this test doesn't need to be HIP, invoking the driver here seems unnecessarily complicated when you can just do `clang -cc1 -triple amdgcn-amd-amdhsa` which doesn't require you do use `__device__` stuff.

jhuber6 reviewed (commented):
So this is just an IR lowering? We'd probably need some more conclusive tests that it works as expected for exceptional values, but I'm going to guess it was copied from an existing implementation so we could just refer to that one.

jhuber6 commented on clang/test/CodeGenHIP/logb_scalbn.cpp:
```diff
@@ -0,0 +1,32 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
```
Move to normal `CodeGen`.

jhuber6 reviewed (commented):
I'll defer to @arsenm for the final verdict.

arsenm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -6011,10 +6011,27 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
   }
   }
 
+  // These will be emitted as Intrinsic later.
+  auto NeedsDeviceOverload = [&](unsigned BuiltinID) {
+    if (getTarget().getTriple().isAMDGCN()) {
+      switch (BuiltinID) {
+      default:
+        return false;
+      case Builtin::BIlogb:
+      case Builtin::BI__builtin_logb:
+      case Builtin::BIscalbn:
+      case Builtin::BI__builtin_scalbn:
+        return true;
+      }
+    }
+    return false;
+  };
```
Should not be a lambda, and should not hardcode the triple. and needs documentation, and a fixme to remove it. Can we check if the libcall is supported or something? Really we should have libm

choikwa replied:
There is `llvm::isLibFuncEmittable(const Module *M, const TargetLibraryInfo *TLI, LibFunc TheLibFunc)` but unsure how to access TLI from CGBuiltin's CodeGenFunction::EmitBuiltinExpr.

arsenm commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -1876,6 +1897,30 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
+  case Builtin::BIlogb:
+  case Builtin::BI__builtin_logb: {
+    auto *Src0 = EmitScalarExpr(E->getArg(0));
+    auto *FrExpFunc = CGM.getIntrinsic(Intrinsic::amdgcn_frexp_exp,
+                                       {Builder.getInt32Ty(), Src0->getType()});
```
Should use the generic frexp intrinsic

arsenm commented on clang/test/CodeGen/logb_scalbn.c:
```diff
@@ -0,0 +1,29 @@
+// CHECK-NEXT:    store float [[CONV1]], ptr [[D2_ASCAST]], align 4
+// CHECK-NEXT:    ret void
+//
+void my_kernel(){
+  float D1 = __builtin_logb((float)16);
+  float D2 = __builtin_scalbn((float)16, 10);
+}
```
Not a kernel. also test one case per function, and test the float and double versions. also with and without constant argument

choikwa commented:
Updated and addressed feedback except the triplet check. At the moment, I'm unsure if there is better way than to query using llvm::isLibFuncEmittable, but that requires TLI.

arsenm commented on clang/test/CodeGen/logb_scalbn.c:
```diff
@@ -0,0 +1,237 @@
+// CHECK-NEXT:    store float [[CONV2]], ptr [[D1_ASCAST]], align 4
+// CHECK-NEXT:    ret void
+//
+void test_logb_var(float a) {
+  float D1 = __builtin_logb(a);
+}
```
Testing the implicit cast isn't useful, the flat versions should test the f suffixed versions

arsenm commented on clang/test/CodeGen/logb_scalbn.c:
```diff
@@ -0,0 +1,237 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
+// RUN: %clang -cc1 -triple amdgcn-amd-amdhsa -emit-llvm -o - %s | FileCheck %s
```
test with fp exceptions and errno too

arsenm commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -1142,6 +1163,31 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
+    auto *FCmpOEQ =
+        Builder.CreateFCmpOEQ(Src0, ConstantFP::getZero(Builder.getDoubleTy()));
+    auto *Sel2 = Builder.CreateSelect(
+        FCmpOEQ, ConstantFP::getInfinity(Builder.getDoubleTy(), /*Neg*/ true),
```
```suggestion
        FCmpOEQ, ConstantFP::getInfinity(Builder.getDoubleTy(), /*Negative=*/true),
```

arsenm commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -1142,6 +1163,31 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
   case AMDGPU::BI__builtin_amdgcn_s_prefetch_data:
     return emitBuiltinWithOneOverloadedType<2>(
         *this, E, Intrinsic::amdgcn_s_prefetch_data);
+  case Builtin::BIlogb:
+  case Builtin::BI__builtin_logb: {
```
logbf?

arsenm commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -1142,6 +1163,31 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
         *this, E, Intrinsic::amdgcn_s_prefetch_data);
+  case Builtin::BIlogb:
+  case Builtin::BI__builtin_logb: {
+    auto *Src0 = EmitScalarExpr(E->getArg(0));
```
no auto throughout

yxsamliu commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -43,6 +43,33 @@ using namespace clang;
+  if (!CGF.CGM.getLangOpts().MathErrno &&
+      CGF.CurFPFeatures.getExceptionMode() ==
+          LangOptions::FPExceptionModeKind::FPE_Ignore &&
+      CGF.getTarget().getTriple().isAMDGCN()) {
```
Instead of checking triple, you can add a virtual function TargetCodeGenInfo::supportsLibCall(), by default it returns true. For AMDGPU, it returns false.

choikwa commented:
Gentle ping

yxsamliu commented:
can we add a llvm-test-suite test for these builtin functions? we can compare their value with host results for typical and corner inputs, like the test added in this PR https://github.com/llvm/llvm-test-suite/pull/230 . This will give us some confidence that the clang codegen works.

choikwa commented:
Created https://github.com/llvm/llvm-test-suite/pull/240

choikwa commented:
@yxsamliu should this be merged before test?
